pkg/k8s/controllers/reconciler.go

```diff
@@ -142,48 +134,22 @@ func (r *Reconciler) endReconcile(
 		return
 	}
 
-	if rr.Skipped {
-		log.Info("Reconciler skipped this event")
-		return
+	if !rr.SkipStatusUpdate {
+		log.Info("Reconciler has applied the CR successfully")
 	}
-
-	log.Info("Reconciler has applied the CR successfully")
 }
 
 // Reconcile is invoked when controller receives resource spec.
 func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	log := r.log.WithField("resourceName", req.NamespacedName).WithField("kind", r.Kind())
 	rr := r.doReconcile(ctx, log, req)
 
-	// defer as func to avoid 'capturing' rr - we want to provide latest values to the endReconcile
-	defer func() {
-		r.endReconcile(ctx, log, *rr)
-	}()
+	r.endReconcile(ctx, log, *rr)
 
-	if rr.ReconcileErr != nil {
-		if rr.PropagateErr {
-			// Controlelr infra will retry immediately.
-			return rr.ControllerRes, rr.ReconcileErr
-		}
-
-		// Note: if CR has been deleted and Handler.NotFound returns an error, we want to retry cause Handler probably
-		// tried to do something (send email, or cleanup or whatever action it takes on Deleted CR) - and failed.
-		// If handler is a no-op on Delete, than no error and we won't retry.
-
-		if rr.Skipped {
-			// no further action needed. Even if failed - honor Reconcile decision.
-			return rr.ControllerRes, nil
-		}
-
-		// if we do not report error to the controller, we need to requeue for this CR
-		if rr.ControllerRes.Requeue || rr.ControllerRes.RequeueAfter > 0 {
-			// Reconcile alredy asked for requeue, honor it
-			return rr.ControllerRes, nil
-		}
-
-		// Reconcile reported an error and did not ask for a requeue.
-		// This is OK, we will requeue with incrementing intervals.
-		rr.ControllerRes.RequeueAfter = requeueDuration(rr.failCount)
+	propagateErr, ok := rr.ReconcileErr.(PropagateError)
```

Comment:
Please do not cast errors to concrete types.  Use [errors.Is](https://pkg.go.dev/errors#Is).  See [this](https://docs.google.com/document/d/1IsO9CfMpkwEJRKRmexvjKtb8XlOVXR_Ja8l99OG4JtM/edit#heading=h.8swsdcpfjiyf).

Reply:
good point, I will apply it to the new repro soon.